plonky3/src/prover/mod.rs

```diff
@@ -0,0 +1,203 @@
+//! A plonky3 prover using FRI and Poseidon
+
+mod types;
+
+use std::io::Read;
+
+use powdr_ast::analyzed::Analyzed;
+
+use powdr_executor::witgen::WitgenCallback;
+
+use p3_air::BaseAir;
+use p3_uni_stark::{prove, verify, Proof};
+use powdr_number::{FieldElement, KnownField};
+use rand::{distributions::Standard, thread_rng, Rng};
+
+use crate::circuit_builder::{cast_to_goldilocks, PowdrCircuit, Val};
+
+use self::types::*;
+
+/// In the context of plonky3, "setup" generation is the process of instantiating Poseidon with randomly sampled constants
+pub fn generate_setup() -> Vec<Val> {
+    let num_rounds = 2 * HALF_NUM_FULL_ROUNDS + NUM_PARTIAL_ROUNDS;
+    let num_constants = WIDTH * num_rounds;
+
+    thread_rng()
+        .sample_iter(Standard)
+        .take(num_constants)
+        .collect()
+}
+
+#[derive(Clone)]
+pub struct Plonky3Prover<'a, T> {
+    /// The analyzed PIL
+    analyzed: &'a Analyzed<T>,
+    /// The value of the fixed columns
+    fixed: &'a [(String, Vec<T>)],
+    /// The constants which uniquely determine the prover, loosely interpreted as a setup
+    setup: Constants,
+    /// The constants which uniquely determine the prover, loosely interpreted as a verification key
+    vkey: Option<Constants>,
```

Comment:
Is this the same as above? Shouldn't the vkey depend on the circuit as well?

Reply:
good call, the verifier I'm using somehow depends on the air still, hmm

Reply:
Ok I think it's fine as is, when you call `verify` the prover has access to the circuit